server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessageVaultRestoreTask.java

```diff
@@ -0,0 +1,174 @@
+        this.userToRestore = userToRestore;
+        this.mailboxManager = mailboxManager;
+        this.restoreMailbox = MailboxPath.forUser(userToRestore.asString(), RESTORE_MAILBOX_NAME);
+        this.successfulRestoreCount = new AtomicLong();
+        this.errorRestoreCount = new AtomicLong();
+        this.failRemovingCount = new AtomicLong();
+    }
+
+    @Override
+    public Result run() {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(userToRestore.asString());
+            MessageManager restoreMailboxManager = restoreMailboxManager(session);
+
+            return Flux.from(deletedMessageVault.search(userToRestore, Query.ALL)).toStream()
+                .map(deletedMessage -> moveToMailbox(restoreMailboxManager, deletedMessage, session))
+                .reduce(Result.COMPLETED, Task::combine);
+        } catch (MailboxException e) {
+            LOGGER.error("Error happens when while creating messageManager for mailbox {}", restoreMailbox.asString(), e);
+            return Result.PARTIAL;
+        }
+    }
+
+    private Result moveToMailbox(MessageManager restoreMailboxManager, DeletedMessage deletedMessage, MailboxSession session) {
+        Result appended = appendToMailbox(restoreMailboxManager, deletedMessage, session);
+
+        return Optional.of(appended)
+            .filter(appendedResult -> appendedResult.equals(Result.COMPLETED))
+            .map(appendedResult -> Task.combine(appendedResult, deleteInTheVault(deletedMessage)))
+            .orElse(Result.PARTIAL);
+    }
+
+    private Result deleteInTheVault(DeletedMessage deletedMessage) {
+        try {
+            Mono.from(deletedMessageVault.delete(userToRestore, deletedMessage.getMessageId())).block();
+            return Result.COMPLETED;
+        } catch (Exception anyException) {
+            LOGGER.error("deleting message {} in the vault got error", deletedMessage.getMessageId().serialize(), anyException);
+            this.failRemovingCount.incrementAndGet();
+            return Result.PARTIAL;
+        }
+    }
+
+    private Result appendToMailbox(MessageManager restoreMailboxManager, DeletedMessage deletedMessage, MailboxSession session) {
+        try {
+            restoreMailboxManager.appendMessage(AppendCommand.builder().build(deletedMessage.getContent().get()), session);
+            this.successfulRestoreCount.incrementAndGet();
+            return Result.COMPLETED;
+        } catch (MailboxException e) {
+            LOGGER.error("append message {} to restore mailbox {} didn't success",
+                deletedMessage.getMessageId().serialize(), restoreMailbox.asString(), e);
+
+            this.errorRestoreCount.incrementAndGet();
+            return Result.PARTIAL;
+        }
+    }
+
+    private MessageManager restoreMailboxManager(MailboxSession session) throws MailboxException {
+        if (mailboxManager.mailboxExists(restoreMailbox, session)) {
```

Comment:
you can't rely on such tests because by the time you retrieve the mailbox, it can be deleted.

Reply:
if not check existing then get or create, so, what is a proper way to do it?

Reply:
you can get then create